warehouse/query-core/src/main/java/datawave/query/tracking/ActiveQueryLog.java

```diff
@@ -0,0 +1,198 @@
+            if (maxIdle > 0) {
+                Cache<String,ActiveQuery> newCache = setupCache(maxIdle);
+                if (this.CACHE == null) {
+                    this.CACHE = newCache;
+                } else {
+                    Cache<String,ActiveQuery> oldCache = this.CACHE;
+                    this.CACHE = newCache;
+                    this.CACHE.putAll(oldCache.asMap());
+                }
+            } else {
+                LOG.error("Bad value: (" + maxIdle + ") for maxIdle");
+            }
+        }
+    }
+    
+    synchronized private void checkSettings(AccumuloConfiguration conf, boolean useDefaults) {
+        
+        String maxIdleStr = conf.get(MAX_IDLE);
+        if (maxIdleStr != null) {
+            try {
+                setMaxIdle(Long.valueOf(maxIdleStr));
+            } catch (NumberFormatException e) {
+                LOG.error("Bad value: (" + maxIdleStr + ") in " + MAX_IDLE + " : " + e.getMessage());
+            }
+        } else if (useDefaults) {
+            setMaxIdle(this.maxIdle);
+        }
+        
+        String logPeriodStr = conf.get(LOG_PERIOD);
+        if (logPeriodStr != null) {
+            try {
+                setLogPeriod(Long.valueOf(logPeriodStr));
+            } catch (NumberFormatException e) {
+                LOG.error("Bad value: (" + logPeriodStr + ") in " + LOG_PERIOD + " : " + e.getMessage());
+            }
+        } else if (useDefaults) {
+            setLogPeriod(this.logPeriod);
+        }
+        
+        String logMaxQueriesStr = conf.get(LOG_MAX_QUERIES);
+        if (logMaxQueriesStr != null) {
+            try {
+                setLogMaxQueries(Integer.valueOf(logMaxQueriesStr));
+            } catch (NumberFormatException e) {
+                LOG.error("Bad value: (" + logMaxQueriesStr + ") in " + LOG_MAX_QUERIES + " : " + e.getMessage());
+            }
+        }
+        
+        String windowSizeStr = conf.get(WINDOW_SIZE);
+        if (windowSizeStr != null) {
+            try {
+                setWindowSize(Integer.valueOf(windowSizeStr));
+            } catch (NumberFormatException e) {
+                LOG.error("Bad value: (" + windowSizeStr + ") in " + WINDOW_SIZE + " : " + e.getMessage());
+            }
+        }
+    }
+    
+    private Cache<String,ActiveQuery> setupCache(long maxIdle) {
+        Caffeine<Object,Object> caffeine = Caffeine.newBuilder();
+        caffeine.expireAfterAccess(maxIdle, TimeUnit.MILLISECONDS);
+        return caffeine.build();
+    }
+    
+    synchronized public void remove(String queryId, Range range) {
+        ActiveQuery activeQuery = get(queryId);
+        int numActiveRanges = activeQuery.removeRange(range);
+        if (numActiveRanges == 0) {
+            this.CACHE.invalidate(queryId);
+        }
+    }
+    
+    synchronized public ActiveQuery get(String queryId) {
+        return this.CACHE.get(queryId, s -> new ActiveQuery(queryId, this.windowSize));
```

Comment:
Is there any way to avoid this method from being synchronized.  Optimally we only synchronize when we need to add a new ActiveQuery (i.e on write).

Reply:
CACHE operations are threadsafe per the API.  The syncronization was due to the possibility of replacing the CACHE object itself when setMaxIdle is changed.  I removed the syncronized on get(), remove(), and setMaxIdle() and used a ReentrandReadWriteLock when using the CACHE object (excpet for the null check in setMaxIdle() which is re-checked inside the locked section anyway.